### ISSUE TITLE
Accessibility for blind and visually impaired screen reader users

### ISSUE THREAD

**Issue by mjanusauskas** [codemirror/CodeMirror#4604]
The known issue of lack of compatibility with screen readers and assistive technology completely prevents blind and visually impaired screen reader users from being able to use and participate in the enormous number of online tools and resources that use CodeMirror as their editor.
It would be wonderful if the need for full inclusion and accessibility support could be given priority by the CodeMirror community.
A working example of how this can be done quite successfully can be found here: 
https://teachaccess.github.io/tutorial/

**Comment by adrianheine** [comment]
I think @marijnh and I are very interested in CodeMirror being accessible and useful to a broad range of people – supporting `contentEditable` [is part of this](https://bugzilla.mozilla.org/show_bug.cgi?id=919711) as far as I know. Without specific funding we will not be able to address broad suggestions such as this, though. If there are specific, easy-to-solve issues, we might be able to fix them on the go.

**Comment by mjanusauskas** [comment]
The severe issue is that screen readers report the contents of the 
editor to be blank and no cursor navigation or reading of the contents 
is possible. The only thing that currently works is pressing CTRL+a to 
select all to hear the editor contents which is not really useful.

It does sound like making the default contenteditable would be a major 
improvement and step in the right direction.

Currently CodeMirror is simply unusable by an unfortunately large number 
of people who depend on accessibility.

Hopefully the contenteditable change will be done soon.

**Comment by schanzer** [comment]
@mjanusauskas this won't be done soon - proper a11y support isn't easy (otherwise everyone would do it!), and the github tutorial doesn't come close to covering what CM will need. I've been doing significant work on a [related project](https://bootstrapworld.github.io/codemirror-blocks/) (and contributed small fixes back into CM) to do this right -- and I can assure you that contentEditable is just the beginning of what's needed.

**Comment by mjanusauskas** [comment]
@schanzer Thank you for the work you have been doing and for contributing back so many more can benefit from your work. While perhaps not easy, it is certain that incorporating accessibility from the beginning of a design is the best solution to avoiding the far more difficult work of trying to add it in later.

**Comment by Neurrone** [comment]
Any estimates on roughly how much funding would be required to fund resolution of this issue?

**Comment by westonruter** [comment]
@adrianheine The WordPress project is looking at integrating CodeMirror into core (see https://github.com/WordPress/codemirror-wp). Given that WordPress puts a high importance on accessibility, there would be interest among the community in funding any needed accessibility improvements as well as funding ongoing maintenance.

**Comment by jasonday** [comment]
> Without specific funding we will not be able to address broad suggestions such as this, though.

 I understand that it will take some work, but choosing to do nothing is actively excluding an entire group of people. The accessibility dev community would likely get involved if an effort was identified.

**Comment by marijnh** [comment]
We are aware of the problems and sympathetic to addressing them. The current architecture of the editor (which is still hugely influenced by the initial 'fake everything because browsers aren't reliable' approach) makes this very hard though. We are considering some radical changes, and with luck, we'll have an announcement about that soon. Pull request for low-hanging improvements are always very welcome (if properly documented so that we understand how hey actually improve things).

**Comment by jasonday** [comment]
@marijnh & @adrianheine - 

Potential approach to establish a baseline level of accessibility for CodeMirror:

Notes:
- I haven't fully tested these approaches, as there are some complex things happening with the textarea within the CodeMirror js
- Screenreader only content [approach](https://webaim.org/techniques/css/invisiblecontent/)

Approach #1 
- Create option for including checkbox to toggle CodeMirror instance for the textarea (formatting only - keeping apis, etc. intact) 
- Or include the checkbox (non-configurable) and hide it offscreen making it available for screenreader users, but not sighted. 

_-Any edit to the textarea or CodeMirror instance should update the other_


Approach #2:
- Screenreader only edit & submit buttons 
    - on edit, user focus is put into textarea which is offscreen
    - on submit, textarea and CodeMirror instance updated
- Remove `display: none;` from original text area
    - `display: none;` removes an element from the accessibility tree

**Comment by schanzer** [comment]
@marijnh , would you be open to sharing some of the radical changes you're considering? I've made significant headway on a project aimed at improving the screen-reader accessibility of _programming in general_, which is built entirely around CodeMirror. I'm both curious about what you're thinking, but also just a *little* worried that any massive API changes will wreak havoc on my codebase.

**Comment by marijnh** [comment]
> would you be open to sharing some of the radical changes you're considering?

We intend to move to contentEditable more completely. There'll be an announcement sometimes in the coming months when we have the design worked out more clearly.

> but also just a little worried that any massive API changes will wreak havoc on my codebase.

Oh yeah, that will definitely happen, and upgrading will take some work.

**Comment by QuincyLarson** [comment]
@marijnh This issue is really important to the freeCodeCamp.org community, which has many members who have visual impairments or are completely blind. Is there anything we can do to help with this?

**Comment by marijnh** [comment]
@QuincyLarson Yes, there'll be work on a more accessible architecture starting in a few months. If you want to contribute financially or be involved in testing, shoot me an [email](marijnh@gmail.com)!

**Comment by tyrylu** [comment]
Any progress on this one?

**Comment by marijnh** [comment]
Work on CodeMirror 6 is progressing in [this repository](https://github.com/codemirror/codemirror.next).

**Comment by tyrylu** [comment]
Thank you, 6.x looks nice from the accessibility perspective.

**Comment by adil-s** [comment]
any update?
